### DISCUSSION TITLE
Paired residue prediction dependencies in ESM2 – Arcadia Science

### DISCUSSION THREAD

**Discussion by giscus** [Arcadia-Science/2024-paired-token-masking#3] [General]
# Paired residue prediction dependencies in ESM2 – Arcadia Science

https://arcadia-science.github.io/2024-paired-token-masking/

**Comment by keithchev** [comment]
My guess is that smaller models are limited by their size to learning local pairwise interactions since those are probably the most "accessible" patterns in the data, and these interactions likely correspond to the contact map. Larger models presumably have the ability to learn longer-range and higher-order interactions, which may either substitute for or dilute out the reliance on explicit pairwise contacts. Indeed the prominent horizontal and vertical bands in the heatmap for the 15B model seem to suggest that the model has shifted to learning longer-range dependencies; it would be interesting to see whether these bands correspond to biologically significant residues (or are instead either uninterpretable or artifacts of some kind).

**Reply by jacobboysen** [reply]
Agreed ^^. This smells like a dimensionality issue? Lots of zooming up and down the manifold happening here. I guess the wet lab takeaway from this work is that smedium models are probably pretty good (maybe preferable?) for designing small peptides, nano bodies, etc but bigger models are needed to design big boi biologics. Context is king, of course. Good read, thanks.

**Reply by ekiefl** [reply]
> This smells like a dimensionality issue? Lots of zooming up and down the manifold happening here

What does zooming up and down the manifold mean?

**Reply by jacobboysen** [reply]
Eh I was being glib but essentially from my read this approach is interrogating ESM2 trained on evolutionary seqs (very high dim) and basically collapsing/summarizing that info into a lower dim, and then comparing to evolution itself (the contact map) that is itself also collapsed into low dim by protein/evo constraints from evolutionary seqs (again, high dim). This is more or less an SAE approach if I’m not mistaken — basically you’ve “found” neuron layers in the smaller model that fire on specific features (here, pairwise constraints, mostly secondary struct). And the larger models aren’t firing on these features at the same magnitude bc their neurons fire stronger on tertiary/globular features and have “selective amnesia” about their secondary bias. Could be wrong, that’s just my read!

**Comment by borgesadair1** [comment]
Thanks for the interesting read! Forgive me if this is a naive question, but could this result be reflecting overfitting on the part of the 15B-parameter model? Like the larger model is always able to provide the "correct" answer via a process more akin to straight up memorization, whereas the smaller model is required to be more efficient, which could be achieved learning more generalizable rules of biology or taking advantage of information shared between residues?

**Reply by ekiefl** [reply]
I'm so glad you questioned this, too, Adair. The hypothesis I like best is that phylogenetic bias in the training data (aka overrepresentation of certain species and genes) leads the model to learn correlations that aren't necessarily due to the functional or structural evolutionary constraints found in proteins but instead reflect the training data composition.

This hypothesis is bolstered by some emerging research that explores training dataset bias in the performance of protein language models, like [this one](https://www.biorxiv.org/content/10.1101/2024.10.03.616542v1) and [this one](https://www.biorxiv.org/content/10.1101/2024.03.07.584001v1).

**Reply by jacobboysen** [reply]
hADA is definitely over represented in the train set, that could explain both the dimensionality issue and this bias-v tradeoff idea. Can you repeat analysis with one or a few proteins under represented or OOD? Maybe small ones to limit costs?

**Reply by ekiefl** [reply]
Sounds good @jacobboysen. If you give me a 1-3 proteins that are (1) under-represented in the training dataset, (2) fold into stable proteins, and (3) are <400AA, I'll re-run this analysis and update the pub to show the results!

**Reply by jacobboysen** [reply]
This is an evolutionary novelty question (which would by definition be either underrep or OOD in uniprot) so: what are Arcadia’s favorite in house proteins (that you want to publicize?) :) I’d guess underrepresented foldable proteins would be  things like extremophile proteins, tardigrade proteins, venoms, immune system, reproductive system, etc. OOD proteins would be engineered ones most likely, such as Top7, Ke70, that ankyrin repeat thing, etc. If it’s a training set bias issue you should in theory see very different profiles from these groups as compared to hADA. Good luck!

**Reply by borgesadair1** [reply]
@ekiefl if you are looking for nominees, I would be happy to help. I have a lot of proteins that might fit this bill (venoms, virus, immune). Also I'm not working on them directly anymore, but phage proteins are also criminally underrepresented in all databases and could be interesting. 

My questions are:
- Do you have a way to evaluate a given protein's enrichment in the ESM2 training dataset relative to hADA? 
- Do you need an experimentally determined crystal structure to calculate contact map?

**Reply by ekiefl** [reply]
Thanks @borgesadair1!

1. I don't have an exact method, but it's definitely possible to back out the number of homologs present in the training data, since we know the training data.
2. Good predictions would suffice. Probably the first thing I'd do is try folding it.

**Reply by Dan-Burns** [reply]
I think @borgesadair1 is right about memorization vs generalization.

The 15B model gets the true amino acid(s) correct 83% of the time for the double masked and 91% for the single masked while the 150M model is only 25% and 50% respectively.

On another note, if you use [getcontacts](https://github.com/getcontacts/getcontacts) to generate the contact map, the correlation with the 150M JS divergence is even more apparent.

**Reply by ekiefl** [reply]
Hey @Dan-Burns, would you like to submit a PR with the improved contact map calculation?

**Reply by Dan-Burns** [reply]
@ekiefl, I just submitted a PR - thanks!

**Comment by delalamo** [comment]
Have you looked at if this pattern holds when examining sequences with fewer homologs? The authors of ESM-2 showed in [Fig 1B and 1C of their paper](https://www.biorxiv.org/content/biorxiv/early/2022/12/21/2022.07.20.500902/F1.large.jpg?width=800&height=600&carousel=1) that model's performance is highly dependent on the number homologs seen during training. For sequences with many homologs, for example, there are no observed gains when increasing the size of the ESM-2 beyond ~150M params or so. By contrast, sequences with few homologs do benefit from increases in model size.

**Reply by ekiefl** [reply]
We haven't looked yet, but that's the plan (see above)! Thanks for the reference to those plots.

**Comment by gksr15** [comment]
This was such an interesting read! I didn’t expect the 150M model to show stronger alignment with the 3D contact map than the much larger 15B model, especially since the bigger model performs better overall in reconstruction accuracy.

I’m wondering if you have any thoughts on why the larger models seem to lose that clear structural correspondence? A few possibilities came to mind:  Maybe the larger models are internalizing structure in a more distributed or abstract way that doesn’t show up as clearly in pairwise comparisons like JS divergence. Or could it be that the predictions become so confident (low perplexity) that the probability shifts caused by double-masking are just too subtle to reflect structural dependencies? I also wondered if the model is learning to "ignore" masks more effectively as it scales, reducing the chance that masking one residue affects another.

Would love to hear your take, and whether you’ve explored other ways to capture these dependencies, like using gradients or even attention maps directly?

**Reply by ekiefl** [reply]
Hey @gksr15, thanks for your comment. I think all of these hypotheses are very reasonable.

> whether you’ve explored other ways to capture these dependencies, like using gradients or even attention maps directly?

What you see is what you get. We're not actively pushing this front, but we think understanding how performance scales with models is really important, and this observation is definitely one lens through which to explore the problem.

**Reply by gwbyeon-uw** [reply]
> whether you’ve explored other ways to capture these dependencies, like using gradients or even attention maps directly?

Another unsupervised extraction of pairwise weights from PLMs here:
https://doi.org/10.1073/pnas.2406285121
https://github.com/zzhangzzhang/pLMs-interpretability

I think Fig. 4A/E and Fig. 5 may provide some insights here?

**Comment by glarue** [comment]
Forgive me if I've missed something, but it seems like we are taking it as read that the smaller model correlates better with the contact map—I would be interested to see this correspondence more quantitatively evaluated. In other words, is it possible that the 15B model contains more or less the same underlying structure (with respect to the contact map) as the 150M model, but with (potentially) additional (longer-range) predicted pairwise dependencies on top?

**Reply by ekiefl** [reply]
> is it possible that the 15B model contains more or less the same underlying structure (with respect to the contact map) as the 150M model

In terms of magnitude, it doesn't appear to be the case--the other dependencies overwhelm the signal. But I suppose it's possible that the contact map structure can still be found. To detect this, you'd need a method robust enough to avoid signal hacking (e.g., extracting a symphony from white noise). Incoming collab?

**Reply by glarue** [reply]
Right—to be clear, I agree with the other insightful commenters about the likely proximate causes of the data presented; in particular, I have seen in my own work examples of highly overrepresented (by way of evolutionary conservation) genes biasing the efficacy of supposedly "general" models (e.g. Evo2) which are significantly less effective when run on much more obscure sequences. That said, I don't see any _a priori_ reason to think the 15B model would be more susceptible to overfitting than the 150M model, but that would of course depend entirely on how both models were originally trained.

I do think it would be interesting to try to differentiate between "the 15B model doesn't capture the contact map as well" versus "the 15B model captures the contact map plus a bunch of other (potentially) interesting predicted dependencies that it has sufficient attention/parameter depth to learn"; as you say, however, this would require careful analysis and I don't—off the top of my head—have an obvious method in mind. Definitely interested to hear if others know of a workable approach!

**Comment by shiakim** [comment]
The unexpected results on model parameter scaling are very interesting. 

As others have alluded, I’d like to highlight the implicit nature of the comparison between the contact map and JS divergence, that both are treated here as approximations to the true biological pairwise dependencies. The unexpected observations of ESM2 outputs naturally draw attention to it, but I think another question that is unclear is to what extent the contact map is representative of the totality of pairwise dependencies of amino acid identities in nature, since the goal of the models is to learn relationships within overall biological context, and not only physical proximity for intra-protein residues.

If the 15B model outputs are more accurate, the discrepancy between the two maps may indicate that larger models are capturing higher order information, such as allostery, larger conformational changes, interaction with other proteins, and other phenomena that may not be reflected in the contact map that is utilized as a reference. It would be interesting to see how the correlation of JS vs contact map looks like for proteins from different families, such as juxtaposing results from proteins with highly complex functions and interactions, and those with simpler roles (to the extent known). For the protein explored here specifically, ADA, I’d be curious to know if some kind of residual between the JS divergence and contact map reveal any clues about higher order function (e.g. some kind of normalized and unit-aligned residuals are high for some surface residues that binds DPP4, where binding to another protein is information that would not be covered in an intra-protein contact map). 

If the 15B model outputs are less reflective of true pairwise dependencies than a contact map may be, there could be several possibilities as well. For instance, the model's attention may be becoming diffuse and perhaps overly weighted towards other properties, which could be an artifact of the model architecture or perhaps the preparation of pretraining data (which would be more tractable). I think it would be informative to see if fine-tuning on the protein classes of interest would restore the contact map semblance or if the discrepancies remain. Testing structure-informed models such as ISM (https://www.biorxiv.org/content/10.1101/2024.11.08.622579v1) may be insightful too.
